moz_sql_parser/formatting.py

```diff
@@ -0,0 +1,242 @@
+    """
+    Escape identifiers.
+
+    ANSI uses single quotes, but many databases use back quotes.
+
+    """
+    if not should_quote(identifier):
+        return identifier
+
+    quote = '"' if ansi_quotes else '`'
+    identifier = identifier.replace(quote, 2*quote)
+    return '{0}{1}{2}'.format(quote, identifier, quote)
+
+
+def Operator(op, parentheses=False):
+    op = ' {0} '.format(op)
+    def func(self, json):
+        out = op.join(self.dispatch(v) for v in json)
+        if parentheses:
+            out = '({0})'.format(out)
+        return out
+    return func
+
+
+class Formatter:
+
+    clauses = [
+        'select',
+        'from_',
+        'where',
+        'groupby',
+        'having',
+        'orderby',
+        'limit',
+        'offset',
+    ]
+
+    # simple operators
+    _concat = Operator('||')
+    _mult = Operator('*')
+    _div = Operator('/', parentheses=True)
+    _add = Operator('+')
+    _sub = Operator('-', parentheses=True)
+    _neq = Operator('<>')
+    _gt = Operator('>')
+    _lt = Operator('<')
+    _gte = Operator('>=')
+    _lte = Operator('<=')
+    _eq = Operator('=')
+    _or = Operator('OR')
+    _and = Operator('AND')
+
+    def __init__(self, ansi_quotes=True, should_quote=should_quote):
+        self.ansi_quotes = ansi_quotes
+        self.should_quote = should_quote
+
+    def format(self, json):
+        if 'union' in json:
+            return self.union(json['union'])
+        else:
+            return self.query(json)
+
+    def dispatch(self, json):
+        if isinstance(json, list):
+            return self.delimited_list(json)
+        if isinstance(json, dict):
+            if 'value' in json:
+                return self.value(json)
+            else:
+                return self.op(json)
+        if isinstance(json, string_types):
+            return escape(json, self.ansi_quotes, self.should_quote)
+        
+        return text_type(json)
+
+    def delimited_list(self, json):
+        return ', '.join(self.dispatch(element) for element in json)
+
+    def value(self, json):
+        parts = [self.dispatch(json['value'])]
+        if 'name' in json:
+            parts.extend(['AS', self.dispatch(json['name'])])
+        return ' '.join(parts)
+
+    def op(self, json):
+        if 'on' in json:
+            return self._on(json)
+
+        if len(json) > 1:
+            raise Exception('Operators should have only one key!')
+        key, value = list(json.items())[0]
+
+        # check if the attribute exists, and call the corresponding method;
+        # note that we disallow keys that start with `_` to avoid giving access
+        # to magic methods
+        attr = '_{0}'.format(key)
+        if hasattr(self, attr) and not key.startswith('_'):
+            method = getattr(self, attr)
+            return method(value)
+
+        return '{0}({1})'.format(key.upper(), value)
+
+    def _exists(self, value):
+        return '{0} IS NOT NULL'.format(self.dispatch(value))
```

Comment:
You do the `'{0}'.format(value)` pattern everywhere.  Do you know something I do not?

Reply:
No, it's just habit. :)